logscan/caddy_parser.go

```diff
@@ -0,0 +1,217 @@
+package logscan
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/bmatcuk/doublestar/v4"
+)
+
+// https://caddyserver.com/docs/caddyfile/directives/log
+type CaddyLogEntry struct {
+	Timestamp   interface{} `json:"ts"`
+	Request     Request     `json:"request"`
+	Duration    float64     `json:"duration"`
+	Size_       int         `json:"size"`
+	Status_     int         `json:"status"`
+	RespHeaders Headers     `json:"resp_headers"`
+}
+
+type Request struct {
+	RemoteAddr string  `json:"remote_addr"`
+	Proto      string  `json:"proto"`
+	Method     string  `json:"method"`
+	Host       string  `json:"host"`
+	URI        string  `json:"uri"`
+	Headers    Headers `json:"headers"`
+}
+
+type Headers struct {
+	UserAgent      []string `json:"User-Agent"`
+	Referer        []string `json:"Referer"`
+	ContentType    []string `json:"Content-Type"`
+	XForwardedFor  []string `json:"X-Forwarded-For"`
+	AcceptLanguage []string `json:"Accept-Language"`
+}
+
+type CaddyParser struct {
+	datetime        string
+	excludePatterns []excludePattern
+}
+
+func (p CaddyParser) Parse(line string) (Line, bool, error) {
+	var logEntry CaddyLogEntry
+	err := json.Unmarshal([]byte(line), &logEntry)
+	if err != nil {
+		fmt.Println("Error unmarshalling JSON:", err)
+		return nil, false, err
+	}
+
+	for _, e := range p.excludePatterns {
+		if logEntry.matchesPattern(e) {
+			return nil, true, nil
+		}
+	}
+	return logEntry, false, nil
+}
+
+var _ LineParser = CaddyParser{}
+var _ Line = CaddyLogEntry{}
+
+func (l CaddyLogEntry) Host() string       { return l.Request.Host }
+func (l CaddyLogEntry) RemoteAddr() string { return l.Request.RemoteAddr }
+func (l CaddyLogEntry) Method() string     { return l.Request.Method }
+func (l CaddyLogEntry) HTTP() string       { return l.Request.Proto }
+func (l CaddyLogEntry) Status() int        { return l.Status_ }
+func (l CaddyLogEntry) Size() int          { return l.Size_ }
+func (l CaddyLogEntry) Path() string {
+	u, err := url.ParseRequestURI(l.Request.URI)
+	if err != nil {
+		return ""
+	}
+	return u.Path
+}
+
+func (l CaddyLogEntry) Query() string {
+	u, err := url.Parse(l.Request.URI)
+	if err != nil {
+		return ""
+	}
+	return u.RawQuery
+}
+
+func (l CaddyLogEntry) Timing() time.Duration {
+	// TODO: `Second` should depend on the log format
+	// {seconds, nano, string} where string in {1m32.05s, 6.31ms}
+	return time.Duration(l.Duration * float64(time.Second))
```

Comment:
I wasn't clear on the best way of implementing this

Reply:
I think this can probably be auto-detected based on the type? Otherwise probably need to add a new flag for it or something. Need to look a bit at what the options are here – if you add a (failing) testcase for all possible options then I would have a better idea.